### ISSUE TITLE
Lib 2.0 (in preview) completion status

### ISSUE THREAD

**Issue by passcod** [watchexec/watchexec#205]
Current code was merged to `main`.

See the [blocked-on: lib 2.0](https://github.com/watchexec/watchexec/issues?q=label%3A%22blocked-on%3A+lib+2.0%22) tag for issues that depend on this. Currently:

- #204
- #187
- #139
- #109
- #67
- #149
- #58
- #183
- #174
- #132
- #166
- #144
- #37
- #36

A note that at this point this is a full refactor/rewrite of the library part and while some things _did_ depend on the async aspect, others didn't... but are still dependent on this work.

Goal:
1. refactor the watchexec **library** to use Tokio (incurring a major version bump)
2. **do not** break compatibility for the CLI

- [x] Break out process handling in a separate crate: https://github.com/watchexec/command-group
- [x] Implement the async version of that crate
  - [x] There is an implementation, but it's [wildly unsafe](https://github.com/watchexec/command-group/blob/v1.0.4/src/tokio/windows.rs#L22-L44) due to a Tokio limitation
  - [x] Patched Tokio to avoid this (https://github.com/tokio-rs/tokio/pull/3998)
  - [x] Waiting on Tokio to release with the patch
  - [x] Use the patched Tokio to safen command-group
- [x] Tokio-ize watchexec library:
  - [x] Process handling
  - [x] Signal handling
  - [x] File events (with Notify 5)
  - [x] Main watchexec interface
    - [x] Configuration
    - [x] Handlers for hooks
    - [x] Event sources
    - [x] Action throttling
    - [x] Command tracking
    - [x] Base hooks
  - [x] Documentation first pass
- [x] Draft implementation of CLI in new library
- [x] Full feature set (required for MVP: ⛵ | for linked issues: 🚢 | nice to haves: 🚀)
  - [x] Hooks
    - [x] 🚀 Pre-spawn
    - [x] 🚢 Post-spawn
    - [x] 🚢 ~~Immediately after command halt~~ (Instead: command completion generates a synthetic event)
  - [x] ⛵ Filesystem event metadata
  - [x] ⛵ Signaling compat on windows
  - ~~🚢 Keyboard events?~~ _scoped out_
  - ~~🚀 Detach stdin for process on option?~~ _scoped out_
  - [x] 🚢 Reset screen as additional action (#186)
  - [x] Pathset
    - [x] ⛵ Inclusion
    - [x] 🚢 https://github.com/notify-rs/notify/issues/353
    - ~~🚀 Exclusion~~
    - ~~🚀 Work around pathset errors~~
    - ~~🚢 https://github.com/watchexec/watchexec/issues/190~~
    - ~~🚢 Non-recursive (https://github.com/watchexec/watchexec/issues/174)~~
    - _Pathset expansion with above items left out of scope for now_
  - [x] 🚢 Tagged Filtering (alpha)
    `struct { in: Option<Path>, on: Particle, op: Op, pat: Pattern }`
    - [x] 🚢 Parsing a filter
    - [x] 🚢 Filtering machinery
    - [x] 🚢 Manual/Config filtering
    - [x] 🚀 Non-path-based filtering
    - [x] 🚢 Positive and negative filtering (ignores and !ignores) (#166)
    - [x] 🚀 Other operations
    - [x] 🚢 Extension filter
    - [x] 🚀 Context for ignores
    - [x] 🚢 gitignores (.gitignore, .git/exclude, ~/.gitignore, ~/.config/git/ignore?, .ignore, etc)
      - [x] 🚢 Find project origins
      - [x] 🚀 Find ignore files globally
      - [x] 🚢 Find ignore files for a project
      - [x] 🚀 Build collection of gitignores from [github/gitignore](https://github.com/github/gitignore): https://github.com/watchexec/gitignores
      - [x] 🚀 Find project types (// possibly extract/remove this?)
      - ~~🚀 Map project types to gitignore~~ (left to downstream crate discretion)
      - [x] 🚢 Parse gitignore-style globs into tagged ignores
      - [x] 🚢 Match paths against globs
      - [x] 🚀 Match non-paths against globs
      - [x] 🚢 Find and parse all ignores for a given path/origin
  - [x] ⛵ Globset Filtering (v1 compat)
    - [x] ⛵ Filtering machinery
    - [x] ⛵ Extension filter
    - [x] ⛵ Load from `IgnoreFile`
  - [x] Change reporting
    - [x] ⛵ In environment variables
    - ~~🚀 Some other way that scales better (#212)~~ _scoped out_
- [x] Document every public item
- [x] Release 2.0.0-pre.0
- [x] Tests
  - [x] Tagged ignores
    - [x] Paths with globs
    - [x] Folder ignores ("confusing" v1 behaviour)
    - [x] Ignore scopes (`in_path`)
    - [x] Other matchers
    - [x] Ignore files
    - [x] Parser
  - [x] Globset ignores
  - [x] Change reporting env vars
  - [x] Outcome processing
- [x] TODOs in the code:
  - [x] Signal matching in tagged filterer
  - [x] Process completion matching in tagged filterer
  - [x] Check glob ignore function
  - [x] span tracing in tagged filters
  - [x] span tracing in globset filters
  - [x] Minimal feature set for tokio etc (to reduce effective dependencies)
  - [x] Make `Event` type serialisable (but don't add the derive yet)
- [x] Release 2.0.0-pre.1

### Next steps: _(tracked separately)_

- Implement v1-compatible CLI
- Put tagged ignores behind opt-in via `WATCHEXEC_FILTERER=tagged` + printed alpha warning
- Editing/review pass over documentation
- Create issues for this and above _scoped out_ items
- Polish as the CLI implementation finds stuff
- Implement Cargo Watch CLI
- Update website
- Publish 2.0.0 stable

**Comment by passcod** [comment]
Status update: took a pause while I was looking for a new job, was successful, so starting hacking here again. Timeline is to get a watchexec (and cargo-watch) v1 CLI powered by Lib 2.0 before the end of the year. Then 2022 will be the year of cool features.
